generator/java.stoneg.py

```diff
@@ -1706,6 +1735,81 @@ def update_by_reference(data_type, namespace):
 
         return visibility
 
+    @staticmethod
+    def _get_data_types_from_doc_ref(api, doc, namespace_context):
+        """
+        Given a documentation string, parse it and return all references to other
+        data types. If there are references to routes, include also the data types of
+        those routes.
+        Args:
+        - doc: The documentation string to parse.
+        - namespace_context: The namespace name relative to this documentation.
+        Returns:
+        - a list of referenced data types
+        """
+        output = []
+        data_types, routes_by_ns = JavaApi._get_data_types_and_routes_from_doc_ref(
+            api, doc, namespace_context)
+        for d in data_types:
+            output.append(d)
+        for ns_name, routes in routes_by_ns.items():
+            ns = api.namespaces[ns_name]
+            for r in routes:
+                for d in ns.get_route_io_data_types_for_route(r):
+                    output.append(d)
+        return output
+
+    @staticmethod
+    def _get_data_types_and_routes_from_doc_ref(api, doc, namespace_context):
+        """
+        Given a documentation string, parse it and return all references to other
+        data types and routes.
+        Args:
+        - doc: The documentation string to parse.
+        - namespace_context: The namespace name relative to this documentation.
+        Returns:
+        - a tuple of referenced data types and routes
+        """
+        assert doc is not None
+        data_types = set()
+        routes = defaultdict(set)
+
+        for match in doc_ref_re.finditer(doc):
+            try:
```

Comment:
this is currently a straight duplicate from the two helper methods in the stone repo, except for swallowing the KeyError block. let me know your thoughts here